bloom.go

```diff
@@ -0,0 +1,310 @@
+	return float64(total) / float64(len(b.bits)*8)
+}
+
+// shrinkToSize returns a resized bloom filter with a bit density close to target.
+// This exploits the fact that a test for a probe x in the bloom filter is actually
+// a test for bit x%len, and a bloom filter of size newlen that divides len is easily
+// derived by oring the bits together len/newlen times. This works because
+// x%newlen == x%len%newlen iff newlen divides len, so we can shrink the bloom filter
+// without having the original probes or keys! This functionality lets us construct
+// a bloom filter while only having an upper bound on cardinality, instead of having
+// to have a separate, expensive input-counting phase.
+func (b *bloom) shrinkToSize(target float64) bloom {
+	if target <= 0.0 || target >= 1.0 {
+		return *b
+	}
+
+	// shrinking sets each output bit to the OR-ed together
+	// output of k=`factor` bits that are set with probability
+	// x=`b.load()`. We want to achieve a target load `y`.
+	//
+	// The probability that a bit is set is one minus the probability
+	// that its inputs are all unset-- 1-(1-x)^k. To get k given y,
+	// https://www.wolframalpha.com/input/?i=solve+for+k+in+y%3D1-%281-x%29%5Ek
+	// => k=log(1-y)/log(1-x)
+	factor := len(b.bits)
+	divisor := math.Log(1 - b.load())
+	if divisor != 0 { // avoid divide by zero for empty filter (b.load() is 0)
+		factor = int(math.Log(1-target) / divisor)
+	}
+
+	// We can only shrink the bloom filter to a size that is a factor of the
+	// input size. This is made easier by bloomSizeBase being highly composite.
+	for factor > 0 && len(b.bits)%factor != 0 {
+		factor--
+	}
+
+	if factor <= 1 {
+		return *b
+	}
+	out := bloom{b.hasher, make([]uint8, len(b.bits)/factor)}
+	j := 0
+	for i := 0; i < len(b.bits); i++ {
+		out.bits[j] |= b.bits[i]
+		j++
+		if j >= len(out.bits) {
+			j = 0
+		}
+	}
+
+	return out
+}
+
+func (b bloom) write(w *writer) {
+	// header: serialization version, hasher id
+	w.Write([]byte{1, bloomHasherIds[reflect.ValueOf(b.hasher).Pointer()]})
+	w.Write(b.bits)
+}
+
+func makeBloomFilterFromEncoded(buf []byte) (bloom, error) {
+	b := bloom{}
+	if len(buf) < 2 || buf[0] != 1 {
+		return b, errors.New("invalid bloom filter encoding (wrong size/version)")
+	}
+	if buf[1] <= 0 || int(buf[1]) > len(bloomHashers) {
+		return b, errors.New("invalid bloom filter encoding (unknown hasher type)")
+	}
+	b.hasher = bloomHashers[buf[1]-1]
+	b.bits = buf[2:]
+	return b, nil
+}
+
+// bloomHasherIds maps from function pointers to hash numbers, to allow
+// backwards compatible hash function changes.
+var bloomHasherIds = map[uintptr]byte{
+	reflect.ValueOf(bloomHasherCRCBlocked64B8K3).Pointer(): 1,
+}
+
+// bloomHashers maps from hash identifierss stored in encoded bloom filters to
+// hash functions, to allo backwards compatible hash function evolution.
+var bloomHashers = []bloomHash{
+	bloomHasherCRCBlocked64B8K3,
+}
+
+// The following functions and constants *must not* be changed unless you can prove
```

Comment:
do we maybe have a test that can detect most changes here? ie hardcoding that output of the hashes for a few inputs and ensuring they don't change.

Reply:
added a test to ensure stability